ENVIRONMENT.md
# Fixed Order Exploration Environment

We explain the working of a fixed order environment using a simple example of 5*5 grid. We use the ANSI Renderer to show the different examples. The configuration used for the sample environment is as follows:

```python
    np.random.seed(100)
    render = "ansi"  # "ansi"  # change to "human"
    env_config = dict(
        width=5,
        height=5,
        population_density=1.0,
        vaccine_density=1.0,
        initial_infection_fraction=0.04,
        initial_vaccination_fraction=0,
        prob_infection=0.2,
        prob_agent_movement=0.0,
        disease_planner_config={
            "incubation_period_mu": 0,
            "incubation_period_sigma": 0,
            "recovery_period_mu": 20,
            "recovery_period_sigma": 0,
        },
        vaccine_score_weight=0.5,
        max_simulation_timesteps=20 * 20 * 10,
        early_stopping_patience=20,
        use_renderer=render,  # can be "human", "ansi"
        use_np_model=True,
        toric=False,
        dummy_simulation=False,
        simulation_single_tick=True,
        debug=True,
        seed=0)
```

The fixed order environment is a simple environment where the agent starts from the top left corner and moves in a fixed order to the the botton right. We show the first 3 steps below where the agent moves from location `(0,0)` to `(2,0)`

![](https://i.imgur.com/qljhCfI.png)

Looking at the various agent states, we can see that the grid has `1` infectious agent and remaining `24` are susceptible agents. Our objective is to stop the infection from spreading to the remaining 24 susceptible agent using the least number of vaccines.

If agent vacciantes the cell positions `(0,0) , (2,0) , (1,0),  (1,1) and (2,1)` if the `toric` is `False` or no horizonal or vertical wrapping of environment is allowed.

If `toric` is `True` , then the agent has to vaccinate 3 other additional cell positions `(4,0), (4,1) and (4,2)`

This vaccination policy , also known as `ring vaccination`  vaccinates the agents surrounding the infectious agents thereby preventing any spread of infection.

The ANSI renderer also shows us other useful information like the

* % of agents in each of the states - Susceptible, Infectious, Recovered and Vaccinated
* Number of Simulation Ticks or time steps and the Environment Steps
* Vaccine Budget or vaccines remaining

Now moving ahead to envrionment step 5, we can see at step 4, the agent chooses to vaccinates cell `0,3` , we can see that our vaccine budget reduces by `1` and we have an agent in vaccinated cell. Note that vaccination is only successful if we apply it on the `susceptible agent`, and only then the agent state changes to `vaccinated`.

![](https://i.imgur.com/iYIHpyg.png)

Moving on, the agent moving in a fixed order can cover the entire grid after 25 steps. The `single_simulation_tick` is set to `True`  which means the agent action is freezed and the simulation will move towards its terminal state after a simulation tick occurs. This happens automatically for the fixed order environment once the entire grid is covered by the agent.

![](https://i.imgur.com/gDtjImc.png)

Notice in the 25th step , our random policy has vaccinated `52%` of the agents or `52% * 25 = 13` agents. We can also see that the agents vaccinated are quite far from ideal as in the last step when the simulation completes, many susceptible agents were infected and eventually moved to the recovered state. The vaccines were also wasted on many agents who were far from the infectious agent.

The visualisation helps us understand how effective our policy is working.  We have also desinged normalised metrics to help us understand how `good` or `bad` our policy is against our twin objectives of `Stopping Infection Spread` and `Minimise utilisation of vaccines`.

As explained in the Environment metrics, we designed 3 metrics for this purpose with the 2 metrics for checking against the 2 objectives and the 3rd metric combining the 2 objectives into a single number.

We show the metric values for this random policy below

```python
'normalized_vaccine_wastage': 0.47, 'normalized_susceptible': 0.0, 'normalized_protected': 0.54
```

Good Policy

We show a better policy trained using an RL Agent trained after 35000 envrionment steps, which is able to prevent vacccine spread but does cause some wastage of vaccines in a `toric 5*5 grid*`. We show the last 2 steps for brevity below

![](https://i.imgur.com/tWlTjk0.png)

We can see that the initial infectious agent moves to recovered, but no other agent is infected or has eventually moved to recovered, thereby showing that our vaccination has prevented other agents from getting infected. Howver, we can see that there are some wasted vaccines. Hence, this is not the perfect vaccination strategy.

We show the metric values for this RL policy below

```python
"normalized_vaccine_wastage": 0.38, "normalized_susceptible": 0.63,    "normalized_protected": 1.00
```

Best Policy

We show the perfect ring vaccination policy trained using an RL Agent trained after 100100 envrionment steps. This agent is able to prevent vacccine spread but also ensures no wastage of vaccines in a `toric 5*5 grid*`.

![](https://i.imgur.com/qnJ1zZH.png)

```python
"normalized_vaccine_wastage": 0, "normalized_susceptible": 1.00,    "normalized_protected": 1.00
```

Notice for the toric grid, the infectious agent at cell `(4,1)` is surrounded by vaccines and hence no infection spreads and there is also no wastage of vaccines as only the 8 cells adjacent to the infectious cell is vaccinated.

## Comparing different vaccination strategy

We can compare the performance across different vaccination strategy by looking at the `normalized_susceptible`  metric.  For e.g. looking at the 2 cases below

Agent 1:

```python
"normalized_vaccine_wastage": 0.38, "normalized_susceptible": 0.63,    "normalized_protected": 1.00
```

Agent 2

```python
"normalized_vaccine_wastage": 0.32, "normalized_susceptible": 0.69,    "normalized_protected": 1.00
```

We can see Agent 2 is better as it has a higher `normalized_susceptible` score. It is also easy to see this by just comparing the `normalized_vaccine_wastage` score as the `normalized_protected = 1` for both of them. If the `normalized_protected` score was higher for Agent 1, then it would not be obvious and we can then use the `normalized_susceptible` score to decide on the better agent.

## Is ring vaccination always the best strategy?

We can have cases where ring vaccination is not always the best strategy. One such case happens when the number of vaccinations available is not enough to ring vaccinate.

Lets assume we have a non-toric grid with width,height as 7,5 respectively with total number of agents as 35. There are 2 Infectious agents and we have only 8 vaccines. We also for simplicity assume that the `simulation_single_tick = True`, so the vaccinations happen in the first time step and then the simulation is allowed to run till completion.

![](https://i.imgur.com/VQa47d8.png)

We can see that the number of vaccines required to fully ring vaccinate the 2 infectious agents is 13. If we follow a simple ring vaccination strategy with the available 8 vaccines, we can only ring vaccinate one infectious agent and the infection spreads due to the other agent. If we run the simulation, we end up with the following scenario as shown in the screenshot below for the last 2 steps.

![](https://i.imgur.com/P26glb2.png)

The normalised metrics are as follows

```python
'normalized_vaccine_wastage': -0.25, 'normalized_susceptible': 0.0, 'normalized_protected': 0.24242424242424243
```

Now, if we were to intelligently try to create grid areas seperating different infectious agents. The last 2 steps now becomes

![](https://i.imgur.com/6XaS3Ci.png)


```python
'normalized_vaccine_wastage': -0.3, 'normalized_susceptible': 1.0, 'normalized_protected': 0.8181818181818181
```


If we make the above problem further difficult, by providing only 5 vaccines, then the intelligent vaccination approach is still able to protect many agents, but the ring vaccination performs worse.

Below screenshot shows the last 2 steps for the ring vaccination strategy

![](https://i.imgur.com/5BzRdlj.png)


```python
'normalized_vaccine_wastage': -0.4, 'normalized_susceptible': 0.0, 'normalized_protected': 0.15151515151515152

```

Similarly for a grid vaccination approach, the last 2 steps are

![](https://i.imgur.com/NGuR2Bu.png)


```python
'normalized_vaccine_wastage': -0.4, 'normalized_susceptible': 1.0, 'normalized_protected': 0.7575757575757575
```


We can clearly see that the grid vaccination performs far better than the ring vaccination. It is able to protect more susceptible agents and also conserve vaccines.

This is only a simple scenario involving 2 infection clusters in a simple `7*5`  grid environment and hence it was easy to find that a grid based vaccination strategy is better than a ring vaccination strategy. When the environment becomes much larger with different infection clusters (which is also what happens practically in real life), it becomes difficult to find these grids and vaccinate.

However, we think an intelligent RL (Reinforcement Learning) agent will be able to learn and provide us with optimal solutions than can help guide future vaccination policy and strategy. We have been able  to train a RL agent that was able to learn ring vaccination by providing a local view of an agent with various local radius sizes ranging from the nearest neighbours to a view involving agents seperated by upto 3 cells. These successful experiments provides us with the impetus so that we can in future also successfully train RL agents that can learn various other vaccination strategies in large environments. A RL solution will also have the added benefit of being scalable which is a critical requirement.

## Apply stochasticity to the problem

We notice that our parameters like the`initial_infection_fraction`, `vaccine density`, `prob_infection` are all constant, but in a real life scenario these are uncertain estimates. We can add a stochastic element to these data and that further increases the complexity of the problem.

We think this also makes an RL solution more well suited to a hand designed expert solution.

## Incorporating Vaccine efficiency

Currently we also assume that the vaccination has a `100%` success rate where an vaccinated agent can not be infected. However as typically seen in practice, vaccines have an efficacy rate less than `100%` thereby making the problem of vaccination tricky.

## Adding multiple time steps for vaccine intravention

We have in all previous examples assumed that we only make the vaccination decision once, and then the simulation happens till the terminal condition is reached. If we allow for vaccine intraventions as the simulation progresses through different time steps, the problem gets more complicated where decisions additionally rely on whether to vaccinate ahead of time or to wait and watch and then vaccinate.


rog_rl/env.py
import gym
from gym import spaces, wrappers
from gym.utils import seeding

from enum import Enum
import numpy as np

from rog_rl.agent_state import AgentState
from scipy.signal import convolve2d


class ActionType(Enum):
    STEP = 0
    VACCINATE = 1


class RogRLEnv(gym.Env):
    def __init__(self, config={}):
        # Setup Config
        self.default_config = dict(
            width=50,
            height=50,
            population_density=0.75,
            vaccine_density=0.05,
            initial_infection_fraction=0.1,
            initial_vaccination_fraction=0.05,
            prob_infection=0.2,
            prob_agent_movement=0.0,
            disease_planner_config={
                "incubation_period_mu": 0,
                "incubation_period_sigma": 0,
                "recovery_period_mu": 20,
                "recovery_period_sigma": 0,
            },
            only_count_successful_vaccines=False,
            vaccine_score_weight=-1,
            use_np_model=True,
            max_simulation_timesteps=200,
            early_stopping_patience=20,
            use_renderer=False,  # can be "simple", "ansi"
            toric=True,
            fast_complete_simulation=True,
            simulation_single_tick=False,
            dummy_simulation=False,
            debug=False,
        )
        self.config = {}
        self.config.update(self.default_config)
        self.update_configs(config)

        self.seed(self.config.get("seed"))

        self.dummy_simulation = self.config["dummy_simulation"]
        self.debug = self.config["debug"]

        self.width = self.config["width"]
        self.height = self.config["height"]

        self.use_renderer = self.config["use_renderer"]
        self.vaccine_score_weight = self.config["vaccine_score_weight"]

        self.set_agent_state()
        self.set_action_type()

        self.disease_model = self.set_disease_model()

        self.action_space = self.set_action_space()
        self.observation_space = self.set_observation_space()

        self._model = None
        self.running_score = None
        self.np_random = np.random

        self.last_action = None
        self.last_action_response = None

        assert self.config[
            "use_np_model"
        ], "Non np model is not \
        , use use_np_model: True"

        self.renderer = False

        if self.use_renderer:
            self.initialize_renderer(mode=self.use_renderer)

        self.cumulative_reward = 0

    def update_configs(self, config={}):
        self.config.update(config)

    def seed(self, seed=None):
        self.np_random, seed = seeding.np_random(seed)
        return [seed]

    def set_observation_space(self):
        return spaces.Box(
            low=np.uint8(0),
            high=np.uint8(1),
            shape=(self.width, self.height, len(AgentState)),
        )

    def set_action_space(self):
        return spaces.MultiDiscrete([len(ActionType), self.width, self.height])

    def set_agent_state(self):
        self.agent_state = AgentState

    def set_action_type(self):
        self.action_type = ActionType

    def set_disease_model(self):
        if self.config["use_np_model"]:
            from rog_rl.model_np import DiseaseSimModel
        else:
            from rog_rl.model import DiseaseSimModel
        return DiseaseSimModel

    def set_renderer(self, renderer):
        self.use_renderer = renderer
        if self.use_renderer:
            self.initialize_renderer(mode=self.use_renderer)

    def reset(self):
        # Delete Model if already exists
        if self._model:
            del self._model

        if self.dummy_simulation:
            """
            In dummy simulation mode
            return a randomly sampled observation
            """
            return self.observation_space.sample()

        self._step_count = 0
        width = self.config["width"]
        height = self.config["height"]
        population_density = self.config["population_density"]
        vaccine_density = self.config["vaccine_density"]
        initial_infection_fraction = self.config["initial_infection_fraction"]
        initial_vaccination_fraction = self.config["initial_vaccination_fraction"]
        prob_infection = self.config["prob_infection"]
        prob_agent_movement = self.config["prob_agent_movement"]
        disease_planner_config = self.config["disease_planner_config"]
        max_simulation_timesteps = self.config["max_simulation_timesteps"]
        only_count_successful_vaccines = self.config["only_count_successful_vaccines"]
        early_stopping_patience = self.config["early_stopping_patience"]
        toric = self.config["toric"]
        fast_complete_simulation = self.config["fast_complete_simulation"]

        """
        Seeding Strategy :
            - The env maintains a custom seed/unsseded np.random instance
            accessible at self.np_random

            whenever env.seed() is called, the said np_random instance
            is seeded

            and during every new instantiation of a DiseaseEngine instance,
            it is seeded with a random number sampled from the self.np_random.
        """
        _simulator_instance_seed = self.np_random.randint(4294967296)
        # Instantiate Disease Model
        self._model = self.disease_model(
            width,
            height,
            population_density,
            vaccine_density,
            initial_infection_fraction,
            initial_vaccination_fraction,
            prob_infection,
            prob_agent_movement,
            disease_planner_config,
            max_simulation_timesteps,
            early_stopping_patience,
            only_count_successful_vaccines,
            fast_complete_simulation,
            toric,
            seed=_simulator_instance_seed,
        )

        self.env_reset()
        # Set the max timesteps of an env as the sum of :
        # - max_simulation_timesteps
        # - Number of Vaccines available

        self._max_episode_steps = self.config.get("max_episode_steps", None)
        if self._max_episode_steps is None:
            self._max_episode_steps = (
                self.config["max_simulation_timesteps"] + self._model.n_vaccines
            )

        #         Tick model
        if not self.config["use_np_model"]:
            # Not needed for model_np
            self._model.tick()

        if self.vaccine_score_weight < 0:
            self.running_score = self.get_current_game_score(
                include_vaccine_score=False
            )
        else:
            self.running_score = self.get_current_game_score(include_vaccine_score=True)
        self.cumulative_reward = 0
        observation = self.get_observation()

        # Used for extra metrics
        self.initial_infection_channel = self._model.infection_scheduled_grid.copy()
        self.n_initial_vaccines = self._model.n_vaccines

        return observation

    def env_reset(self):
        # Initialize location of vaccination agent
        pass

    def get_observation(self):
        obs = self._model.get_observation()
        return self.post_process_observation(obs)

    def post_process_observation(self, observation):
        return observation

    def get_action_response(self):
        return self.last_action_response

    def initialize_renderer(self, mode="human"):

        if self.use_renderer in ["simple"]:
            self.metadata = {
                "render.modes": ["simple", "rgb_array"],
                "video.frames_per_second": 5,
            }
            from rog_rl.renderer import SimpleRenderer

            self.renderer = SimpleRenderer(grid_size=(self.width, self.height))

        elif mode in ["human", "rgb_array"]:
            self.metadata = {
                "render.modes": ["human", "rgb_array"],
                "video.frames_per_second": 5,
            }
            from rog_rl.renderer import Renderer

            self.renderer = Renderer(grid_size=(self.width, self.height))
        elif mode in ["ansi"]:
            """
            Initialize ANSI Renderer here
            """
            self.metadata = {
                "render.modes": ["human", "ansi"],
                "video.frames_per_second": 5,
            }
            from rog_rl.renderer import ANSIRenderer

            self.renderer = ANSIRenderer()

        elif mode in ["PIL"]:
            """
            Initialize PIL Headless Renderer here
            for visualising during training
            """
            self.metadata = {
                "render.modes": ["PIL", "rgb_array"],
                "video.frames_per_second": 5,
            }
            from rog_rl.renderer import PILRenderer

            self.renderer = PILRenderer(grid_size=(self.width, self.height))

        else:
            print("Invalid Mode selected for render:", mode)

        self.renderer.setup(mode=mode)

    def get_agents_by_state(self, state):
        if self.config["use_np_model"]:
            obs = self._model.observation
            states = np.argmax(obs, axis=-1)
            idx = np.where(states == state.value)
            return [i for i in zip(idx[0], idx[1])]
        else:
            scheduler = self._model.get_scheduler()
            return scheduler.get_agents_by_state(state)

    def get_agents_grid(self):
        if self.config["use_np_model"]:
            obs = self._model.observation
            return np.argmax(obs, axis=-1)
        else:
            scheduler = self._model.get_scheduler()
            return scheduler.get_agents_by_state()

    def get_agent_positions(self, agent):
        if self.config["use_np_model"]:
            agent_x, agent_y = agent
            return agent_x, agent_y
        else:
            agent_x, agent_y = agent.pos
            return agent_x, agent_y

    def update_renderer(self, mode="human"):
        """
        Updates the latest board state on the renderer
        """
        # Draw Renderer
        # Update Renderer State
        model = self._model
        if self.config["use_np_model"]:
            total_agents = model.n_agents
            _simulation_steps = model.schedule_steps
        else:
            scheduler = model.get_scheduler()
            total_agents = scheduler.get_agent_count()
            _simulation_steps = int(scheduler.steps)

        state_metrics = self.get_current_game_metrics()

        initial_vaccines = int(model.initial_vaccination_fraction * model.n_agents)

        _vaccines_given = model.max_vaccines - model.n_vaccines - initial_vaccines

        # Game Steps includes steps in which each agent is vaccinated
        _game_steps = _simulation_steps + _vaccines_given

        self.renderer.update_stats("SCORE", "{:.3f}".format(self.cumulative_reward))
        self.renderer.update_stats("VACCINE_BUDGET", "{}".format(model.n_vaccines))
        self.renderer.update_stats("SIMULATION_TICKS", "{}".format(_simulation_steps))
        self.renderer.update_stats("GAME_TICKS", "{}".format(_game_steps))
        self.renderer.update_stats("ENV_STEPS", "{}".format(self._step_count))

        self.update_env_renderer_stats()

        if self.use_renderer == "simple":
            for key in state_metrics:
                self.renderer.update_stats(key, state_metrics[key])
            obs = self._model.get_observation()
            return self.renderer.get_render_output(obs)

        elif self.use_renderer == "ansi":
            for key in state_metrics:
                self.renderer.update_stats(key, state_metrics[key])
            grid = self.get_agents_grid()
            render_output = self.renderer.render(self.width, self.height, grid)
            if self.debug:
                print(render_output)
            return render_output

        for _state in AgentState:
            key = "population.{}".format(_state.name)
            stats = state_metrics[key]
            self.renderer.update_stats(
                key, "{} ({:.2f}%)".format(int(stats * total_agents), stats * 100)
            )
            if mode in ["human", "rgb_array"]:
                color = self.renderer.COLOR_MAP.get_color(_state)
                agents = self.get_agents_by_state(_state)
                for _agent in agents:
                    _agent_x, _agent_y = self.get_agent_positions(_agent)
                    self.renderer.draw_cell(_agent_x, _agent_y, color)
        if mode in ["human", "rgb_array"]:
            # Update the rest of the renderer
            self.renderer.pre_render()

            # Only in case of recording via Monitor or setting mode = rgb_array
            # we require the rgb image
            if isinstance(self, wrappers.Monitor):
                return_rgb_array = mode in ["human", "rgb_array"]
            else:
                return_rgb_array = mode == "rgb_array"
            render_output = self.renderer.post_render(return_rgb_array)
            return render_output

    def update_env_renderer_stats(self):
        pass

    def get_current_game_score(self, include_vaccine_score):
        """
        Returns the current game score

        The game score is currently represented as :
            (percentage of susceptibles left in the population)
        """
        score = self._model.get_population_fraction_by_state(AgentState.SUSCEPTIBLE)
        if include_vaccine_score:
            score += self._model.get_population_fraction_by_state(AgentState.VACCINATED)

        return score

    def get_current_game_metrics(self, dummy_simulation=False):
        """
        Returns a dictionary containing important game metrics
        """
        _d = {}
        # current population fraction of different states
        for _state in AgentState:
            if not dummy_simulation:
                _value = self._model.get_population_fraction_by_state(_state)
            else:
                _value = self.np_random.rand()

            _key = "population.{}".format(_state.name)
            _d[_key] = _value
        # Add "Protected" and "Affected"
        _d["population.PROTECTED"] = (
            _d["population.SUSCEPTIBLE"] + _d["population.VACCINATED"]
        )
        _d["population.AFFECTED"] = 1.0 - _d["population.PROTECTED"]
        # Add R0 to the game metrics
        #         _d["R0/10"] = self._model.contact_network.compute_R0()/10.0
        return _d

    def get_end_of_simulation_metrics(self, dummy_simulation=False):
        """
        Returns a dictionary of metrics that should only be calculated at end of simuation
        """
        if dummy_simulation:
            return {}

        stats = {}
        # current population fraction of different states
        for _state in AgentState:
            _value = self._model.get_population_fraction_by_state(_state)
            _key = "population.{}".format(_state.name)
            stats[_key] = _value

        protected = stats["population.SUSCEPTIBLE"] + stats["population.VACCINATED"]

        _d = {}
        n_infected = np.sum(self.initial_infection_channel)
        k = np.ones((3, 3))
        # This will include vaccination and infeted
        ring_vaccination = (
            convolve2d(
                self.initial_infection_channel, k, "same", boundary=self._model.boundary
            )
            > 0
        )
        agent_present = np.bitwise_not(self._model.no_agent_grid)
        n_agents = self._model.n_agents
        ring_vaccination = np.bitwise_and(ring_vaccination, agent_present)
        min_vaccines_needed = np.sum(ring_vaccination) - n_infected
        max_susceptible_possible = n_agents - np.sum(ring_vaccination)
        max_vaccines_possible = n_agents - n_infected
        vaccines_used = self.n_initial_vaccines - max(0, self._model.n_vaccines)
        if min_vaccines_needed < max_vaccines_possible:
            _d["normalized_vaccine_wastage"] = (vaccines_used - min_vaccines_needed) / (
                max_vaccines_possible - min_vaccines_needed
            )
        else:
            # Weird edge case where everyone needs to be vaccinated
            _d["normalized_vaccine_wastage"] = (
                float(vaccines_used == min_vaccines_needed) - 1.0
            )
        if max_susceptible_possible > 0:
            _d["normalized_susceptible"] = (
                stats["population.SUSCEPTIBLE"] * n_agents / max_susceptible_possible
            )
        else:
            _d["normalized_susceptible"] = 1.0
        _d["normalized_protected"] = protected / (1 - n_infected / n_agents)

        return _d

    def calculate_rewards(self):
        # Compute difference in game score
        if self.vaccine_score_weight < 0:
            current_score = self.get_current_game_score(include_vaccine_score=False)
            _step_reward = current_score - self.running_score
            self.cumulative_reward += _step_reward
            self.running_score = current_score
        else:
            current_score = self.get_current_game_score(include_vaccine_score=True)
            _step_reward = current_score - self.running_score
            self.running_score = current_score
            _done = not self._model.is_running()
            if _done:
                _step_reward = self.terminal_reward(current_score, _step_reward)
            self.cumulative_reward += _step_reward

        return _step_reward

    def terminal_reward(self, current_score, _step_reward):
        susecptible_percentage = self.get_current_game_score(
            include_vaccine_score=False
        )
        _step_reward -= (
            current_score - susecptible_percentage
        ) * self.vaccine_score_weight
        return _step_reward

    def step_action(self, action):

        _observation = False

        response = "STEP"

        return _observation, response

    def step_tick(self):
        """
        Handle SIM_TICK action
        """
        # Handle action propagation in real simulator
        if not self.config.get("simulation_single_tick", False):
            self._model.tick()
        else:
            self._model.run_simulation_to_end()
        _observation = self._model.get_observation()
        response = "STEP"
        return _observation, response

    def step(self, action):
        # Handle dummy_simulation Mode
        if self.dummy_simulation:
            return self.dummy_env_step()

        assert self.action_space.contains(action), "%r (%s) invalid" % (
            action,
            type(action),
        )
        if self._model is None:
            raise Exception("env.step() called before calling env.reset()")

        _done = False
        _info = {}

        _observation, response = self.step_action(action)
        if self._step_count >= self._max_episode_steps:
            self._model.run_simulation_to_end()
            _observation = self._model.get_observation()
        self._step_count += 1

        self.last_action = action
        self.last_action_response = response

        _step_reward = self.calculate_rewards()

        # Add custom game metrics to info key
        game_metrics = self.get_current_game_metrics()
        for _key in game_metrics.keys():
            _info[_key] = game_metrics[_key]

        _info["cumulative_reward"] = self.cumulative_reward
        _done = not self._model.is_running()
        if _done:
            end_sim_metrics = self.get_end_of_simulation_metrics()
            for _key in end_sim_metrics.keys():
                _info[_key] = end_sim_metrics[_key]

        _observation = self.post_process_observation(_observation)
        return _observation, _step_reward, _done, _info

    def dummy_env_step(self):
        """
        Implements a fake env.step for faster Integration Testing
        with RL experiments framework
        """
        observation = self.observation_space.sample()
        reward = self.np_random.rand()
        done = True if self.np_random.rand() < 0.01 else False
        info = {}
        game_metrics = self.get_current_game_metrics(dummy_simulation=True)
        info.update(game_metrics)

        return observation, reward, done, info

    def render(self, mode="human"):
        """
        This methods provides the option to render the
        environment's behavior to a window which should be
        readable to the human eye if mode is set to 'human'.
        """
        if not self.use_renderer:
            return

        if not self.renderer:
            self.initialize_renderer(mode=mode)

        return self.update_renderer(mode=mode)

    def close(self):
        if self.renderer:
            if hasattr(self.renderer, "close"):
                self.renderer.close()
            self.renderer = False
        if self._model:
            # Delete the model instance if it exists
            self._model = None


if __name__ == "__main__":

    render = "ansi"  # "ansi"  # change to "human"
    env_config = dict(
        width=5,
        height=7,
        population_density=1.0,
        vaccine_density=1.0,
        initial_infection_fraction=0.04,
        initial_vaccination_fraction=0,
        prob_infection=0.2,
        prob_agent_movement=0.0,
        disease_planner_config={
            "incubation_period_mu": 0,
            "incubation_period_sigma": 0,
            "recovery_period_mu": 20,
            "recovery_period_sigma": 0,
        },
        max_simulation_timesteps=200,
        early_stopping_patience=20,
        use_renderer=render,
        use_model_np=True,
        fast_complete_simuation=True,
        toric=False,
        dummy_simulation=False,
        debug=True,
    )
    env = RogRLEnv(config=env_config)
    print("USE RENDERER ?", env.use_renderer)
    record = False
    if record:
        # records the the rendering in the `recording` folder
        env = wrappers.Monitor(env, "recording", force=True)

    observation = env.reset()
    done = False
    k = 0

    if not record:
        env.render(mode=render)
    while not done:
        _action = env.action_space.sample()
        # _action = input("Enter action - ex: [1, 4, 2] : ")
        # if _action.strip() == "":
        #     _action = env.action_space.sample()
        # else:
        #     _action = [int(x) for x in _action.split()]
        #     assert _action[0] in [0, 1]
        #     assert _action[1] in list(range(env._model.width))
        #     assert _action[2] in list(range(env._model.height))
        print("Action : ", _action)
        observation, reward, done, info = env.step(_action)
        if not record:
            env.render(mode=render)
        k += 1
    print(info)
    # print(observation.shape)
    # print(k, reward, done)
    # print(observation.shape())
